### PR TITLE
[PERF] further drop changes

### PR DESCRIPTION
**Pull request by matthewjasper**
r? @ghost

### CODE REVIEW

File: src/librustc/ty/instance.rs [hunk 1]
```diff
@@ -181,8 +181,13 @@ impl<'tcx> InstanceDef<'tcx> {
             // drops of `Option::None` before LTO. We also respect the intent of
             // `#[inline]` on `Drop::drop` implementations.
             return ty.ty_adt_def()
-                .and_then(|adt_def| adt_def.destructor(tcx))
-                .map_or(ty.is_enum(), |dtor| tcx.codegen_fn_attrs(dtor.did).requests_inline());
+                .map_or(true, |adt_def| {
+                    adt_def.destructor(tcx)
+                        .map_or(
```
Comment (lzutao) [inline src/librustc/ty/instance.rs]: Maybe using lazy `map_or_else` is better.

File: src/librustc_mir/borrow_check/mod.rs [hunk 2]
```diff
@@ -632,6 +632,7 @@ impl<'cx, 'tcx> DataflowResultsConsumer<'cx, 'tcx> for MirBorrowckCtxt<'cx, 'tcx
             }
             TerminatorKind::Drop {
                 location: ref drop_place,
+                flag: _,
                 target: _,
                 unwind: _,
```
Comment (lzutao) [inline src/librustc_mir/borrow_check/mod.rs]: Maybe using `..` in this and other places is better.

**Comment by matthewjasper** [comment]
That's all for now.

**Comment by matthewjasper** [comment]
I'll do some more tests now that #67332 has merged.

File: src/librustc_mir/interpret/terminator.rs [hunk 3]
```diff
@@ -100,14 +101,22 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                if should_drop {
+                    let instance = Instance::resolve_drop_in_place(*self.tcx, ty);
+                    self.drop_in_place(
+                        place,
+                        instance,
+                        terminator.source_info.span,
+                        target,
+                        unwind
+                    )?;
+                }
```
Comment (RalfJung) [inline src/librustc_mir/interpret/terminator.rs]: I assume this should jump to `target` even if it doesn't drop?
Then you'll need an `else` branch that calls `go_to_block` (like e.g. `Assert`).

File: src/librustc/mir/mod.rs [hunk 4]
```diff
@@ -1080,8 +1080,18 @@ pub enum TerminatorKind<'tcx> {
...
-    Drop { location: Place<'tcx>, target: BasicBlock, unwind: Option<BasicBlock> },
+    /// Drop the `Place`, possibly conditioned on a flag being true.
+    Drop {
+        location: Place<'tcx>,
+        /// Whether to drop the value.
+        ///
+        /// Before drop elaboration this is always `None. After drop elaboration
+        /// If this is `None` then the drop is unconditional, otherwise the drop
+        /// is only evaluated when the flag is true.
+        flag: Option<Place<'tcx>>,
```
Comment (RalfJung) [inline src/librustc/mir/mod.rs]: Why is this a place, when all we do is read from it (so `Operand` would be enough)?

**Comment by Dylan-DPC** [comment]
@matthewjasper any updates?

**Comment by matthewjasper** [comment]
Not really, closing for now.

**Pull request closed without merging**